### PR TITLE
[feat] check NaNs in losses during training, and exit when it happens

### PR DESCRIPTION
**Pull request by ronghanghu**
Add config option `training.exit_on_nan_losses` (Default to True).

When `training.exit_on_nan_losses` is True, check for NaNs in the loss dict and exit the training when NaN happens.

Test plan: tested locally; a new test case `test_exit_on_nan_losses` is added.

### CODE REVIEW

File: mmf/trainers/core/training_loop.py [hunk 1]
```diff
@@ -164,10 +164,25 @@ def run_training_epoch(self) -> None:
 
     def run_training_batch(self, batch: Dict[str, Tensor], loss_divisor: int) -> None:
         report = self._forward(batch)
+        if self.training_config.exit_on_nan_losses:
+            self._check_nan_losses(report)
         loss = extract_loss(report, loss_divisor)
```
Comment (apsdehal) [inline mmf/trainers/core/training_loop.py:169]: You can check afterwards on the loss as well, no? That would anyways be nan. I think ideal place for this check would be just before actual backward happens.

File: mmf/trainers/core/training_loop.py [hunk 2]
```diff
@@ -164,10 +164,25 @@ def run_training_epoch(self) -> None:
 
     def run_training_batch(self, batch: Dict[str, Tensor], loss_divisor: int) -> None:
         report = self._forward(batch)
+        if self.training_config.exit_on_nan_losses:
```
Comment (apsdehal) [inline mmf/trainers/core/training_loop.py:167]: Probably add a simple test for this following existing trainer tests if possible?

**Review by apsdehal — CHANGES_REQUESTED** [review]

File: mmf/trainers/core/training_loop.py (hunk 1 above)
Comment (ronghanghu) [reply mmf/trainers/core/training_loop.py:169]: I was deliberately putting it over `report` rather than loss and before `extract_loss`, so that it would check for each loss before reduction, and provide some clue on which loss and which process lead to NaNs. But I can change it to just check the output of `extract_loss` if you feel that's better 😄

File: mmf/trainers/core/training_loop.py (hunk 2 above)
Comment (ronghanghu) [reply mmf/trainers/core/training_loop.py:167]: Sounds good!

**Comment by ronghanghu** [comment]
A new test case `test_exit_on_nan_losses` is added.

**Comment by vedanuj** [comment]
Looks good. Can you import to phabricator and run the internal tests. Thanks!

**Review by apsdehal — APPROVED** [review]
Please rebase before landing.
